File: src/teste_integracao_com_c/clog.zig
// Wrapper Zig Idiomático
const std = @import("std");
const c = @cImport({
    @cInclude("clog.h");
});

pub const Level = enum(c_int) {
    debug = c.LOG_DEBUG,
    info = c.LOG_INFO,
    warn = c.LOG_WARN,
    err = c.LOG_ERROR,
};

pub const Logger = struct {
    raw: *c.Logger,
    name: []const u8,

    const Self = @This();

    pub fn new(name: []const u8, min_level: Level) !Self {
        const c_name: [*c]const u8 = @ptrCast(name.ptr);
        // name já é [:0]const u8, então name.ptr coerge diretamente para [*c]const u8
        const raw = c.logger_new(c_name, @intCast(@intFromEnum(min_level)));

        if (raw == null) return error.OutOfMemory;

        return Self {
            .raw = raw.?,
            .name = name,
        };
    }

    pub fn deinit(self: Self) void {
        c.logger_free(self.raw);
    }

    pub fn log(self: Self, level: Level, comptime fmt: []const u8, args: anytype) void {
        var buf: [1024]u8 = undefined;
        // Concatena com \x00 no comptime para garantir o terminador nulo para o C
        const message = std.fmt.bufPrint(&buf, fmt ++ "\x00", args) catch |errr| {
            std.debug.print("Erro ao formatar log: {}\n", .{errr});
            return;
        };

        const c_msg: [*c]const u8 = @ptrCast(message.ptr);
        c.logger_log(self.raw, @intCast(@intFromEnum(level)), c_msg);
    }

    pub fn debug(self: Self, comptime fmt: []const u8, args: anytype) void {
        self.log(.debug, fmt, args);
    }

    pub fn info(self: Self, comptime fmt: []const u8, args: anytype) void {
        self.log(.info, fmt, args);
    }

    pub fn warn(self: Self, comptime fmt: []const u8, args: anytype) void {
        self.log(.warn, fmt, args);
    }

    pub fn err(self: Self, comptime fmt: []const u8, args: anytype) void {
        self.log(.err, fmt, args);
    }

    pub fn setLevel(self: Self, level: Level) void {
        c.logger_set_level(self.raw, @intCast(@intFromEnum(level)));
    }
};

File: src/teste_integracao_com_c/clog.c
#include "clog.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Logger {
    char* name;
    LogLevel min_level;
};

Logger* logger_new(const char* name, LogLevel min_level) {
    Logger* logger = (Logger*)malloc(sizeof(Logger));
    if (!logger) return NULL;

    logger->name = (char*)malloc(strlen(name) + 1);
    if (!logger->name) {
        free(logger);
        return NULL;
    }
    strcpy(logger->name, name);

    logger->min_level = min_level;
    return logger;
}

void logger_free(Logger* logger) {
    if (logger) {
        free(logger->name);
        free(logger);
    }
}

void logger_log(Logger* logger, LogLevel level, const char* message) {
    if (!logger) return;
    if (level < logger->min_level) return;

    const char* level_str = "UNKNOWN";
    switch (level) {
        case LOG_DEBUG: level_str = "DEBUG"; break;
        case LOG_INFO:  level_str = "INFO";  break;
        case LOG_WARN:  level_str = "WARN";  break;
        case LOG_ERROR: level_str = "ERROR"; break;
    }

    printf("[%s] (%s): %s\n", level_str, logger->name, message);
}

void logger_set_level(Logger* logger, LogLevel level) {
    if (logger) {
        logger->min_level = level;
    }
}
